src/write_sensu.c

```diff
@@ -0,0 +1,1230 @@
+		}
+	}
+	if (status != 0) {
+		sensu_free(host);
+		return status;
+	}
+
+	if (host->metrics && (host->metric_handlers.nb_strs == 0)) {
+			sensu_free(host);
+			WARNING("write_sensu plugin: metrics enabled but no MetricHandler defined. Giving up.");
+			return -1;
+		}
+
+	if (host->notifications && (host->notification_handlers.nb_strs == 0)) {
+		sensu_free(host);
+		WARNING("write_sensu plugin: notifications enabled but no NotificationHandler defined. Giving up.");
+		return -1;
+	}
+
+	if ((host->notification_handlers.nb_strs > 0) && (host->notifications == 0)) {
+		WARNING("write_sensu plugin: NotificationHandler given so forcing notifications to be enabled");
+		host->notifications = 1;
+	}
+
+	if ((host->metric_handlers.nb_strs > 0) && (host->metrics == 0)) {
+		WARNING("write_sensu plugin: MetricHandler given so forcing metrics to be enabled");
+		host->metrics = 1;
+	}
+
+	if (!(host->notifications || host->metrics)) {
+		WARNING("write_sensu plugin: neither metrics nor notifications enabled. Giving up.");
+		sensu_free(host);
+		return -1;
+	}
+
+	ssnprintf(callback_name, sizeof(callback_name), "write_sensu/%s", host->name);
+	ud.data = host;
+	ud.free_func = sensu_free;
+
+	pthread_mutex_lock(&host->lock);
+
+	if (host->metrics) {
+		status = plugin_register_write(callback_name, sensu_write, &ud);
+		if (status != 0)
+			WARNING("write_sensu plugin: plugin_register_write (\"%s\") "
+					"failed with status %i.",
+					callback_name, status);
+		else /* success */
+			host->reference_count++;
+	}
+
+	if (host->notifications) {
+		status = plugin_register_notification(callback_name, sensu_notification, &ud);
+		if (status != 0)
+			WARNING("write_sensu plugin: plugin_register_notification (\"%s\") "
+					"failed with status %i.",
+					callback_name, status);
+		else
+			host->reference_count++;
+	}
+
+	if (host->reference_count <= 1) {
+		/* Both callbacks failed => free memory.
+		 * We need to unlock here, because sensu_free() will lock.
+		 * This is not a race condition, because we're the only one
+		 * holding a reference. */
+		pthread_mutex_unlock(&host->lock);
+		sensu_free(host);
+		return -1;
+	}
+
+	host->reference_count--;
+	pthread_mutex_unlock(&host->lock);
+
+	return status;
+} /* }}} int sensu_config_node */
+
+static int sensu_config(oconfig_item_t *ci) /* {{{ */
+{
+	int		 i;
+	oconfig_item_t	*child;
+	int		 status;
+	struct str_list sensu_tags_arr;
+
+	sensu_tags_arr.nb_strs = 0;
+	sensu_tags_arr.strs = NULL;
+	sensu_tags = malloc(sizeof(char));
+	if (sensu_tags == NULL) {
+		ERROR(alloc_err);
+		return -1;
+	}
+	sensu_tags[0] = '\0';
+
+	for (i = 0; i < ci->children_num; i++)  {
+		child = &ci->children[i];
+
+		if (strcasecmp("Node", child->key) == 0) {
+			sensu_config_node(child);
+		} else if (strcasecmp(child->key, "attribute") == 0) {
+			char *key = NULL;
+			char *val = NULL;
+
+			if (child->values_num != 2) {
+				WARNING("sensu attributes need both a key and a value.");
```

Comment:
I ran the build with clang's "scan-build" static analysis tool, and I got this warning:
`Potential leak of memory pointed to by 'sensu_tags_arr.strs'`

Not sure it's worth caring in this case, what do you think ?